designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,504 @@
+        "line":1,
+        "column":1,
+        "nodeType":"Identifier",
+        "suppressions":[
+          {
+            "kind":"directive",
+            "justification":"Sample justification message."
+          }
+        ],
+        "messageId":"undef",
+        "endLine":1,
+        "endColumn":2
+      }
+    ],
+    "errorCount":0,
+    "warningCount":0,
+    "fixableErrorCount":0,
+    "fixableWarningCount":0,
+    "usedDeprecatedRules":[]
+  }
+]
+```
+
+### Add new CLI option `--track-suppressions`
+
+We propose to add a new CLI option `--track-suppressions` for the new feature. Without the option, ESLint will work as it is currently, and `suppressions` property should not exist in the output. With this option, ESLint is expected to export all violations, including suppressed ones. `suppressions` should be _empty_ for non-suppressed violations, while `suppressions` should be a list for suppressed ones, according to [suppressions property](https://docs.oasis-open.org/sarif/sarif/v2.0/csprd02/sarif-v2.0-csprd02.html#_Toc10127852).
+
+### Parse description in comments as `justification`
+
+We propose not to ignore the text after two or more dashes in comments and considered it as the justification of a suppression. The modification should be in `getDirectiveComments`.
+
+### Add suppressions attribute in `LintMessage`
+
+`LintMessage` is the type that finally output to the formatter. Currently it contains `ruleId`, `severity`, `message`, etc. What we desire is a `suppressions` attribute in `LintMessage` when the CLI option `--track-suppressions` used. It should be a list that contains a suppression entry (`{kind: "directive", justification: "Fake justification message."}`).
+
+### Reserve external suppression
+
+In the step of apply external suppression (or disabling rules inside of configuration files), ESLint will not skip rules disabled in configuration files if it gets the `--track-suppressions` option. Violations (or problems in ESLint) will be created with `{kind: "external", justification: "Globally disabled"}` if the related rules are switched to `off`.
+
+More specifically, when traversing `configuredRules` in `runRules`, ESLint will not [skip disabled rules](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/linter.js#L893) if the `--track-suppressions` option used. Then the suppression info `{kind: "external", justification: "Globally disabled"}` will be put into [`createReportTranslator`](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/linter.js#L924) and [`createProblem`](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/report-translator.js#L357) to create reports with suppressions for disabled rules. If a non-existing rule is switched to `off`, ESLint would keep the current behavior and would not throw errors. The non-existing rules would not be recorded in suppression information as well.
```

Comment:
Attempting to run rules that have been externally disabled could be tricky:

1. Some shared configs, most notably `eslint-config-airbnb`, explicitly configure every rule, including those it sets to `off`. Would this report those rules as suppressed even though they were never really on in the first place?
2. It's not safe to assume that every rule can run on any given file. For example, trying to run a core rule on a TypeScript file could crash. How would this know which externally-suppressed rules to run?

Reply:
Re 1, `@ljharb/eslint-config` does the same.

Re 2, there's some rules that only work on HTML or MD or JSX or TS files, etc, and they do indeed crash (as they should) when ran on the wrong type of file.

There's also rules that are *supposed* to warn on production files, but are supposed to *not* warn on test files - an app might handle this with configuration overrides. Should this be reported on?

Reply:
I think the intent is that whatever ends up in the final calculated config for a file is what is reported. That would take into account overrides and such.

Reply:
That’s true, though I suppose people who want to use this will understand that tradeoff to get the suppression information. 

@Yiwei-Ding do you agree? You would most likely get about 300 rules disabled in any sufficiently complex project, any turning them on to get the suppression info would also dramatically slow down linting.

Reply:
Sorry for the delay. We didn't realize your complex scenarios :joy:.

Again, we don't want to break any existing pipelines, so in my new update, we propose to add a new severity `suppress` to mark rules as suppressed ones. Without `--track-suppressions`, it just works as `off`. While with `--track-suppressions`, ESLint will run the suppressed rules and record suppression info, and will not report errors/warnings.

Is it good to you?

Reply:
This doesn’t really change the problem, we are just adding more complexity by needing to support a new severity level.

At this point, I don’t think tracking suppressions from config files really makes sense for ESLint. I understand it might make sense with other tools but it seems the nature of ESLint configs is sufficiently complex that I don’t think you’re getting useful data from config-file suppressions. There’s just no way to know that’s it’s the user’s intent vs. a config they are inheriting, even with a new severity.

I think tracking in-line suppressions makes a lot more sense, as both the intent is clear and there’s space to add a reason.

Reply:
I’ve done this on many very large codebases (tried to tighten linting, with many tens of thousands of violations at first).

I’ve never avoided having to manually, or programmatically with a script, add explicit overrides for each violation/file combo (either using inline override comments, or once the feature became available, config file overrides), and then gradually over time, removing old files from the override and fixing them. Time-based expiry just creates a “time bomb” where builds will start failing in the future with no apparent cause and with no guarantee it’s a convenient time to fix them.